# miARma #

miARma is a fully customizable pipeline for NGS transcriptome analyses. Including gene/transcripts, miRNAs and circRNAs expression measurements.
Created at Computational Biology and Bioinformatics Group (CbBio)
Institute of Biomedicine of Seville. IBIS (Spain)
Copyright (c) 2016 IBIS. All rights reserved.

### 1. Included in miARma ###

#### Quality Software ####
* Fastqc v0.10.1
#### Trimming Softare ####
* CutAdapt v13-100
* Minion v13-100
* Reaper v13-100
#### Aligners ####
* Bowtie v1.0.0
* Bowtie v2.2.0
* TopHat v2.0.13
* BWA v0.7.12
* Entity Quantification
* feactureCounts v1.4.6
* Ciri v1.2
* miRDeep v2
#### Others ####
* RNAfold v2.1.9
* Samtools v1.2

### 2. Pre-requisites ###

miARma-Seq is a tool that provides an easy and common interface to various analysis software. It also intends to reduce to the minimum the number of dependencies. Nevertheless, some basic programs listed below must be correctly installed:

* Perl v5.6.0 or higher.
* Java JDK v.1.6. or higher.
* R environment v.3.2 or higher.
* Bioconductor v.1.3 or higher.
#### Compilers: #####
* Apple:
** Xcode
* Linux:
** Gcc
** make

### How do I get set up? ###

* Summary of set up
* Configuration
* Dependencies
* Database configuration
* How to run tests
* Deployment instructions

### Contribution guidelines ###

* Writing tests
* Code review
* Other guidelines

### Who do I talk to? ###

* miARma-devel@cbbio.es

